We're going on an Adventure!

JRoy opened the pull request:
https://user-images.githubusercontent.com/10731363/147530817-f9adc58e-18a5-49ed-84c6-106e51d6948f.mp4

Closes #2029
Closes #2391

## Todo
- [x] Per Player Locale
  - [x] Add support for them
  - [x] Finish porting everything over to per-player-locale system.
- [x] MiniMessage Support (Components - Hover/Click Events)
	- No messages will be converted to MiniMessage in this PR, a separate PR will add this.
	- Currently, there are a few messages which are a giant pain in the ass to convert to support components. They will be converted on a need-to basis.

JRoy committed:
Fix bstats *still* being transitive

JRoy committed:
Add support for per-player-locale and MiniMessage

JRoy committed:
Begin per-player-locale conversion

zml2008 commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:
```diff
@@ -625,44 +636,44 @@ private void runVersion(final Server server, final CommandSource sender, final S
         switch (supportStatus) {
             case NMS_CLEANROOM:
-                sender.sendMessage(ChatColor.DARK_RED + tl("serverUnsupportedCleanroom"));
+                sender.sendComponent(ess, sender.tlComponent(ess, "serverUnsupportedCleanroom").color(AdventureUtil.asAdventure(ChatColor.DARK_RED)));
```
why not just NamedTextColor.DARK_RED?

JRoy committed:
Prevent all locales from falling back to English

JRoy committed:
zml woke up

JRoy committed:
Bump adventure, moar debug

JRoy committed:
Bump run-paper

JRoy committed:
Fix audiences not working

Turns out that adventure-platform registers events which are required to
create an audience for them. Move the creation of the BukkitAudiences
object after we unregister all of our events.

JRoy committed:
i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this i hate this

JRoy committed:
get out of my head get out of my head get out of my head get out of my head get out of my head get out of my head get out of my head get out of my head

kashike commented on Essentials/src/main/java/com/earth2me/essentials/CommandSource.java:
```diff
@@ -21,6 +27,46 @@ public final Player getPlayer() {
+        }
+
+        final String translation = tlLiteral(tlKey, args);
+        if (!translation.startsWith("MM||")) {
```
Should shove `MM||` into a constant, this isn't the only line it is referenced/checked.

JRoy committed:
Move MiniMessage prefix to constant

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Actually support TranslatableException in EssentialsCommand#showError

JRoy committed:
Give CommandSource an Essentials instance on creation

JRoy committed:
Bump checkstyle

JRoy committed:
i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die i want to die

JRoy committed:
Revert "Bump checkstyle"

This reverts commit 90103ba5991f8d2a5c705fd517a7d0336028a968.

JRoy committed:
some progress or whatever

JRoy committed:
Use adventure HSVLike in DownsampleUtil

JRoy committed:
The main module compiles

JRoy committed:
AntiBuild converted

JRoy committed:
Chat converted

JRoy committed:
Discord converted

JRoy committed:
GeoIP converted

JRoy committed:
Protect converted

JRoy committed:
Spawn converted

JRoy committed:
XMPP converted lmfao

JRoy committed:
tests converted

JRoy committed:
begone old method

JRoy commented:
Per player locale and MiniMessage support has been completed and this pull request is now ready for review :)

JRoy committed:
Fix missing check for broadcasting afk messages

JRoy committed:
Don't send messages which yield an empty translation

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Fix compile errors

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/CommandSource.java:
```diff
@@ -21,7 +29,47 @@ public final Player getPlayer() {
+        }
+
+        final String translation = tlLiteral(tlKey, args);
+        if (!translation.startsWith(I18n.MINI_MESSAGE_PREFIX)) {
```
Dunno how I feel about having a prefix for MiniMessage support. Why not just a configuration option?

JRoy replied:
Discussed internally

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/Essentials.java:
```diff
@@ -1128,6 +1152,53 @@ private int broadcastMessage(final IUser sender, final String permission, final
+            }
+
+            final Object[] processedArgs;
+            // FUCK THIS STUPID BULLSHIT
```
```suggestion
```

JRoy replied:
i find this comment to be useful for future maintainers

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/User.java:
```diff
@@ -49,8 +53,10 @@
+import static com.earth2me.essentials.I18n.tlLiteral;
+import static com.earth2me.essentials.I18n.tlLocale;
 
+@SuppressWarnings("deprecation")
```
Should suppress for specific calls, not for the whole class, since this would hide future instances of deprecation that we might actually want to address.

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandbanip.java:
```diff
@@ -48,21 +48,21 @@ public void run(final Server server, final CommandSource sender, final String co
         }
-
-        ess.broadcastMessage("essentials.banip.notify", tl("playerBanIpAddress", senderDisplayName, ipAddress, banReason));
+        
```
```suggestion

```

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/Essentials.java
#	Essentials/src/main/java/com/earth2me/essentials/IEssentials.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandnick.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandrepair.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandrest.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandworth.java
#	Essentials/src/main/java/com/earth2me/essentials/signs/SignTrade.java

JRoy committed:
Adventure don't rename method for more than a day challenge [IMPOSSIBLE]

JRoy committed:
Remove class deprecation suppression for User

JRoy committed:
Remove some whitespace

zml2008 commented on Essentials/src/main/java/com/earth2me/essentials/I18n.java:
```diff
@@ -150,7 +150,7 @@ public String format(final Locale locale, final String string, final Object... o
         final Object[] processedArgs;
         if (miniMessage) {
-            processedArgs = mutateArgs(objects, s -> MiniMessage.miniMessage().escapeTokens(s));
+            processedArgs = mutateArgs(objects, s -> MiniMessage.miniMessage().escapeTags(s));
```
-SNAPSHOT moment

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/User.java
#	EssentialsChat/src/main/java/com/earth2me/essentials/chat/EssentialsChatPlayerListenerNormal.java

JRoy committed:
Adventure 4.10.0

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	build-logic/build.gradle.kts

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandnick.java

JRoy committed:
Update adventure

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Fix build

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/Backup.java
#	Essentials/src/main/java/com/earth2me/essentials/Essentials.java
#	Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java
#	Essentials/src/main/java/com/earth2me/essentials/EssentialsUpgrade.java
#	Essentials/src/main/java/com/earth2me/essentials/I18n.java
#	Essentials/src/main/java/com/earth2me/essentials/Jails.java
#	Essentials/src/main/java/com/earth2me/essentials/ManagedFile.java
#	Essentials/src/main/java/com/earth2me/essentials/Mob.java
#	Essentials/src/main/java/com/earth2me/essentials/Settings.java
#	Essentials/src/main/java/com/earth2me/essentials/Warps.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandsell.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/EssentialsCommand.java
#	Essentials/src/main/java/com/earth2me/essentials/config/EssentialsConfiguration.java
#	Essentials/src/main/java/com/earth2me/essentials/textreader/HelpInput.java
#	EssentialsAntiBuild/src/main/java/com/earth2me/essentials/antibuild/EssentialsConnect.java
#	EssentialsDiscord/src/main/java/net/essentialsx/discord/EssentialsDiscord.java
#	EssentialsGeoIP/src/main/java/com/earth2me/essentials/geoip/EssentialsGeoIPPlayerListener.java
#	EssentialsProtect/src/main/java/com/earth2me/essentials/protect/EssentialsConnect.java

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Update adventure

JRoy committed:
Merge remote-tracking branch 'upstream/refactor/we-are-going-on-an-adventure' into refactor/we-are-going-on-an-adventure

mdcfe committed:
Fix usage of non-initialised logger in handleCrash

mdcfe committed:
Prevent relocation of Paper native adventure usage

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy commented:
TODO minimessage support for mails. Gonna need to move the mail line logic out of the service (or change its behavior :/)

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Use the kezz magic character so we're allowed to escape tags

JRoy committed:
Add support for minimessage in mails

JRoy committed:
Fix hex color support in mini message

JRoy committed:
Fix some oops

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	EssentialsChat/src/main/java/com/earth2me/essentials/chat/EssentialsChatPlayerListenerNormal.java

JRoy committed:
Update latest to adventure

JRoy committed:
lmfao

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Fix nullability issues

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandban.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandbanip.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandkick.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandmute.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandtempban.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandtempbanip.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandtogglejail.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandunban.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandunbanip.java

JRoy committed:
Load resource bundles off the main thread

There are I/O incurred here, players could rapidly swap languages
which could end up leading to the server timing out all players.

While the bundle is being loaded, the defaultBundle is used. In
normal I/O conditions, this should only affect one message.

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandseen.java

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/Kit.java
#	Essentials/src/main/java/com/earth2me/essentials/MailServiceImpl.java
#	Essentials/src/main/java/com/earth2me/essentials/Trade.java
#	Essentials/src/main/java/com/earth2me/essentials/User.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandbook.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandclearinventory.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandgive.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandhat.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commanditem.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commanditemlore.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commanditemname.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandskull.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandunlimited.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandwhois.java

JRoy committed:
Update link module to adventure

LcyDev commented:
I'm praying for your sanity man

codebycam commented:
LGTM

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandhat.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandnuke.java
#	EssentialsDiscord/src/main/java/net/essentialsx/discord/interactions/InteractionControllerImpl.java

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/Console.java:
```diff
@@ -63,6 +65,21 @@ public void sendMessage(final String message) {
+            sendMessage(translation);
+            return;
+        }
+        ((Essentials) ess).getBukkitAudience().sender(getCommandSender()).sendMessage(MiniMessage.miniMessage().deserialize(translation.substring(AdventureUtil.MINI_MESSAGE_PREFIX.length())));
```
This is a very long line

mdcfe commented on providers/1_8Provider/src/main/java/net/ess3/provider/providers/LegacyPlayerLocaleProvider.java:
```diff
@@ -0,0 +1,16 @@
+public class LegacyPlayerLocaleProvider implements PlayerLocaleProvider {
+    @Override
+    public String getLocale(Player player) {
+        return player.spigot().getLocale();
```
This will fail on CraftBukkit. I don't see much point in using NMS, but we may want to have a fallback that returns a fixed value or null.

JRoy replied:
I do not want to have a fallback, no. CraftBukkit users are not real.

pop4959 replied:
There's no reason not to have a fallback there, regardless. Not a big fan of `Player.Spigot` usage.

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/Trade.java:
```diff
@@ -193,23 +191,23 @@ public void isAffordableFor(final IUser user, final CompletableFuture<Boolean> f
 
         if (getMoney() != null && getMoney().signum() > 0 && !user.canAfford(getMoney())) {
-            future.completeExceptionally(new ChargeException(tl("notEnoughMoney", NumberUtil.displayCurrency(getMoney(), ess))));
+            future.completeExceptionally(new ChargeException(user.playerTl("notEnoughMoney", NumberUtil.displayCurrency(getMoney(), ess))));
```
We now have `TranslatableException`. Do we similarly want `ChargeException`s to also take in the key and be translated at display time?

JRoy replied:
TranslatableException is made for running within the command system, and futures are scary here and I don't want to mix things up.

mdcfe reviewed (commented):
Preliminary review

JRoy committed:
md has a small monitor

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/net/ess3/api/events/LocalChatSpyEvent.java
#	EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java

JRoy committed:
with this commit we fix compatability with a staggering 1% of users

mdcfe committed:
Bump adventure to 4.12.0 and platform to 4.2.0

mdcfe committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

mdcfe committed:
Update adventure for 1.19.4

JRoy committed:
Fix wrong broadcast method being used in /sucicide

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Fix build

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commanddelhome.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandeditsign.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandrenamehome.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandsethome.java

JRoy committed:
Fix build?????????????????????

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java
#	build-logic/build.gradle.kts

JRoy committed:
Fix build

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandhome.java
#	EssentialsDiscord/src/main/java/net/essentialsx/discord/JDADiscordService.java
#	EssentialsDiscord/src/main/java/net/essentialsx/discord/commands/Commanddiscordbroadcast.java
#	EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/listeners/LinkBukkitListener.java
#	settings.gradle.kts

mdcfe commented on providers/PaperProvider/build.gradle:
```diff
@@ -18,3 +19,7 @@ essentials {
 }
+
+shadowJar {
+    relocate 'net.kyori.adventure', 'net.essentialsx.temp.adventure'
```
Wasn't this already done for #4941?

JRoy replied:
no

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Fix build

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/I18n.java:
```diff
@@ -207,5 +288,13 @@ public ResourceBundle newBundle(final String baseName, final Locale locale, fina
+
+        @Override
+        public Locale getFallbackLocale(String baseName, Locale locale) {
+            if (baseName == null || locale == null) {
+                throw new NullPointerException();
+            }
+            return null;
```
```suggestion
            if (baseName == null || locale == null) {
                throw new NullPointerException();
            }
            return null;
```
This will break fallbacks from `en_US` to `en`, no?

JRoy replied:
no, works correctly

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandafk.java:
```diff
@@ -50,55 +49,63 @@ private void toggleAfk(final User sender, final User user, final String message)
                 final String dateDiff = sender.getMuteTimeout() > 0 ? DateUtil.formatDateDiff(sender.getMuteTimeout()) : null;
                 if (dateDiff == null) {
-                    throw new Exception(sender.hasMuteReason() ? tl("voiceSilencedReason", sender.getMuteReason()) : tl("voiceSilenced"));
+                    if (sender.hasMuteReason()) {
+                        throw new TranslatableException("voiceSilencedReason", sender.getMuteReason());
+                    } else {
+                        throw new TranslatableException("voiceSilenced");
+                    }
+                }
+                if (sender.hasMuteReason()) {
+                    throw new TranslatableException("voiceSilencedReasonTime", dateDiff, sender.getMuteReason());
+                } else {
+                    throw new TranslatableException("voiceSilencedTime", dateDiff);
                 }
-                throw new Exception(sender.hasMuteReason() ? tl("voiceSilencedReasonTime", dateDiff, sender.getMuteReason()) : tl("voiceSilencedTime", dateDiff));
             }
```
This code has been copied around a few places, would it be worth pulling out into something like `User::getMutedException`?

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/utils/AdventureUtil.java:
```diff
@@ -0,0 +1,55 @@
+    private AdventureUtil() {
+    }
+
+    public static Component toComponent(final String text) {
+        return LEGACY_SERIALIZER.deserialize(text);
+    }
```
I'm fine with having a legacy parser util method but why is this called `toComponent`?

mdcfe commented on EssentialsAntiBuild/src/main/java/com/earth2me/essentials/antibuild/EssentialsConnect.java:
```diff
@@ -30,14 +30,13 @@ IEssentials getEssentials() {
-    void alert(final User user, final String item, final String type) {
+    void alert(final User user, final String item, final String tlKey) {
         final Location loc = user.getLocation();
-        final String warnMessage = tl("alertFormat", user.getName(), type, item, loc.getWorld().getName() + "," + loc.getBlockX() + "," + loc.getBlockY() + "," + loc.getBlockZ());
```
why

mdcfe replied:
wait nevermind

JRoy committed:
toComponent -> deserializeLegacy

JRoy committed:
Initial conversion code, just exists so far

JRoy committed:
Add configurable primary and secondary colors

JRoy committed:
Get rid of the kezz magic char

Kotlin users may not be mentioned in our source code

JRoy committed:
Allow passing in unescaped tags

JRoy committed:
Get rid of the mini message prefix

might as well do this right

JRoy committed:
Hopefully mark all args that need to be parsed

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/ISettings.java
#	Essentials/src/main/java/com/earth2me/essentials/Settings.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandmail.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandskull.java

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/utils/AdventureUtil.java:
```diff
@@ -52,4 +84,36 @@ public static String minifyLegacy(String text) {
+        return ess != null ? ess.getSettings().getSecondaryColor() : Tag.styling(NamedTextColor.RED);
+    }
+
+    public static class Placeholder {
```
This could do with a more descriptive name:
- `ParsedPlaceholder`?
- `MiniMessagePlaceholder`? (feels way too ambiguous)

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/utils/AdventureUtil.java:
```diff
@@ -1,41 +1,73 @@
+        return serializeLegacy(miniMessage().deserialize(format));
+    }
+
     public static String minifyLegacy(String text) {
```
Took a minute to figure out what these all do. Could maybe make the method names more consistent (at least change `minifyLegacy` to `legacyToMini`, not so sure about the `(de)serializeLegacy`)

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/utils/AdventureUtil.java:
```diff
@@ -52,4 +84,36 @@ public static String minifyLegacy(String text) {
+        return new Placeholder(literal);
+    }
+
+    public static Tag supplyTag(final boolean primary) {
+        if (primary) {
+            return ess != null ? ess.getSettings().getPrimaryColor() : Tag.styling(NamedTextColor.GOLD);
+        }
+        return ess != null ? ess.getSettings().getSecondaryColor() : Tag.styling(NamedTextColor.RED);
+    }
```
Perfect opportunity to change up the defaults 👀

JRoy replied:
later, need to spot difference first

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandfly.java:
```diff
@@ -37,9 +38,9 @@ protected void togglePlayer(final CommandSource sender, final User user, Boolean
             }
 
-            user.sendTl("flyMode", user.playerTl(enabled ? "enabled" : "disabled"), user.getDisplayName());
+            user.sendTl("flyMode", AdventureUtil.parsed(user.playerTl(enabled ? "enabled" : "disabled")), user.getDisplayName());
```
I wonder if we could have a util methods for common placeholders like this, instead of pasting `AdventureUtil.parsed(user.playerTl(enabled ? "enabled" : "disabled"))` everywhere

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/EssentialsUpgrade.java:
```diff
@@ -986,6 +988,42 @@ public void generateUidCache() {
+                }
+            }
+        } catch (final Throwable e) {
+            ess.getLogger().log(Level.SEVERE, "Error while upgrade custom locales", e);
```
```suggestion
            ess.getLogger().log(Level.SEVERE, "Error while upgrading custom locales", e);
```

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/EssentialsUpgrade.java:
```diff
@@ -986,6 +988,42 @@ public void generateUidCache() {
         }
     }
 
+    public void upgradeLang() {
```
Probably worth logging a message at the top here

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java:
```diff
@@ -20,25 +21,21 @@ public Commandhelpop() {
-        return message;
+        final String message = FormatUtil.stripFormat(getFinalArg(args, 0));
+        ess.getLogger().log(Level.INFO, tlLiteral("helpOp", from, message));
+        ess.broadcastTl(null, u -> !u.isAuthorized("essentials.helpop.receive") && u.equals(user), "helpOp", from, message);
```
Correct me if I'm wrong but I'm not sure this predicate is doing the same thing as the old behaviour.

Previously we broadcasted to everyone with the `essentials.helpop.receive` permission, and *also* sent it to the sender if they don't have that permission. Now we're broadcasting it, but only excluding the sender and only if they don't have the permission?

Best guess at the intended check:
```suggestion
        ess.broadcastTl(null, u -> !u.isAuthorized("essentials.helpop.receive") && !u.equals(user), "helpOp", from, message);
```

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/signs/SignException.java:
```diff
@@ -1,11 +1,14 @@
+import net.ess3.api.TranslatableException;
+
+public class SignException extends TranslatableException {
+    public SignException(final String message, final Object... args) {
```
```suggestion
    public SignException(final String tlKey, final Object... args) {
```

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/signs/SignException.java:
```diff
@@ -1,11 +1,14 @@
 
-    public SignException(final String message, final Throwable throwable) {
-        super(message, throwable);
+    public SignException(final Throwable cause, final String message, final Object... args) {
```
```suggestion
    public SignException(final Throwable cause, final String tlKey, final Object... args) {
```

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandworth.java:
```diff
@@ -80,15 +79,19 @@ private BigDecimal itemWorth(final CommandSource sender, final User user, final
 
         final BigDecimal result = worth.multiply(BigDecimal.valueOf(amount));
-        sender.sendMessage(MaterialUtil.getDamage(is) != 0 ? tl("worthMeta", is.getType().toString().toLowerCase(Locale.ENGLISH).replace("_", ""), MaterialUtil.getDamage(is), NumberUtil.displayCurrency(result, ess), amount, NumberUtil.displayCurrency(worth, ess)) : tl("worth", is.getType().toString().toLowerCase(Locale.ENGLISH).replace("_", ""), NumberUtil.displayCurrency(result, ess), amount, Nu ...
+        if (MaterialUtil.getDamage(is) != 0) {
+            sender.sendTl("worthMeta", is.getType().toString().toLowerCase(Locale.ENGLISH).replace("_", ""), MaterialUtil.getDamage(is), NumberUtil.displayCurrency(result, ess), amount, NumberUtil.displayCurrency(worth, ess));
+        } else {
+            sender.sendTl("worth", is.getType().toString().toLowerCase(Locale.ENGLISH).replace("_", ""), NumberUtil.displayCurrency(result, ess), amount, NumberUtil.displayCurrency(worth, ess));
+        }
```
```suggestion
        final String itemName = is.getType().toString().toLowerCase(Locale.ENGLISH).replace("_", "");
        if (MaterialUtil.getDamage(is) != 0) {
            sender.sendTl("worthMeta", itemName, MaterialUtil.getDamage(is), NumberUtil.displayCurrency(result, ess), amount, NumberUtil.displayCurrency(worth, ess));
        } else {
            sender.sendTl("worth", itemName, NumberUtil.displayCurrency(result, ess), amount, NumberUtil.displayCurrency(worth, ess));
        }
```
I hope this compiles

Could also possibly separate out `totalCurrency = NumberUtil.displayCurrency(result, ess)` and `perItemCurrency = NumberUtil.displayCurrency(worth, ess)`?

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandkick.java:
```diff
@@ -51,8 +52,10 @@ public void run(final Server server, final CommandSource sender, final String co
         final String senderDisplayName = sender.isPlayer() ? sender.getPlayer().getDisplayName() : Console.DISPLAY_NAME;
 
-        ess.getLogger().log(Level.INFO, tl("playerKicked", senderDisplayName, target.getName(), kickReason));
-        ess.broadcastMessage("essentials.kick.notify", tl("playerKicked", senderDisplayName, target.getName(), kickReason));
+        final String tlKey = "playerKicked";
+        final Object[] objects = {senderDisplayName, target.getName(), kickReason};
+        ess.getLogger().log(Level.INFO, tlLiteral(tlKey, objects));
+        ess.broadcastTl(null, "essentials.kick.notify", tlKey, objects);
```
I'm actually not convinced this is significantly more readable/clear

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandnick.java:
```diff
@@ -39,41 +38,41 @@ public void run(final Server server, final CommandSource sender, final String co
         } else {
             setNickname(server, sender, target, nick);
-            target.sendMessage(tl("nickSet", ess.getSettings().changeDisplayName() ? target.getDisplayName() : nick));
+            target.sendTl("nickSet", ess.getSettings().changeDisplayName() ? target.getDisplayName() : nick);
```
Are we adding MM support to nicknames? Presumably this will be left for a subsequent PR, just want to check

JRoy replied:
i don't know if we want to expose mini message to users or just configs, open to internal discussion

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandrealname.java:
```diff
@@ -31,7 +29,7 @@ protected void run(final Server server, final CommandSource sender, final String
             if (FormatUtil.stripFormat(u.getDisplayName()).toLowerCase(Locale.ENGLISH).contains(lookup)) {
                 foundUser = true;
-                sender.sendMessage(tl("realName", u.getDisplayName(), u.getName()));
+                sender.sendTl("realName", u.getDisplayName(), u.getName());
```
Display names can contain legacy colour codes, what happens here now?

JRoy replied:
fuck

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandsell.java:
```diff
@@ -117,8 +118,8 @@ private BigDecimal sellItem(final User user, final ItemStack is, final String[]
         user.giveMoney(result, null, UserBalanceUpdateEvent.Cause.COMMAND_SELL);
-        user.sendMessage(tl("itemSold", NumberUtil.displayCurrency(result, ess), amount, is.getType().toString().toLowerCase(Locale.ENGLISH), NumberUtil.displayCurrency(worth, ess)));
-        ess.getLogger().log(Level.INFO, tl("itemSoldConsole", user.getName(), is.getType().toString().toLowerCase(Locale.ENGLISH), NumberUtil.displayCurrency(result, ess), amount, NumberUtil.displayCurrency(worth, ess), user.getDisplayName()));
+        user.sendTl("itemSold", NumberUtil.displayCurrency(result, ess), amount, is.getType().toString().toLowerCase(Locale.ENGLISH), NumberUtil.displayCurrency(worth, ess));
+        ess.getLogger().log(Level.INFO, tlLiteral("itemSoldConsole", user.getName(), is.getType().toString().toLowerCase(Locale.ENGLISH), NumberUtil.displayCurrency(result, ess), amount, NumberUtil.displayCurrency(worth, ess), user.getDisplayName()));
```
This sure is a couple of lines

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandtogglejail.java:
```diff
@@ -76,19 +76,29 @@ public void run(final Server server, final CommandSource sender, final String co
-                        final String notifyMessage = timeDiff > 0 ? tl("jailNotifyJailedFor", player.getName(), DateUtil.formatDateDiff(finalDisplayTime), sender.getSender().getName()) : tl("jailNotifyJailed", player.getName(), sender.getSender().getName());
-                        ess.getLogger().log(Level.INFO, notifyMessage);
-                        ess.broadcastMessage("essentials.jail.notify", notifyMessage);
+                        final String tlKey;
+                        final Object[] objects;
+                        if (timeDiff > 0) {
+                            tlKey = "jailNotifyJailedFor";
+                            objects = new Object[]{player.getName(), DateUtil.formatDateDiff(finalDisplayTime)};
+                            sender.sendTl("playerJailedFor", player.getName(), DateUtil.formatDateDiff(finalDisplayTime));
+                        } else {
+                            tlKey = "jailNotifyJailed";
+                            objects = new Object[]{player.getName(), sender.getSender().getName()};
+                            sender.sendTl("playerJailed", player.getName());
+                        }
+
+                        ess.getLogger().log(Level.INFO, tlLiteral(tlKey, objects));
+                        ess.broadcastTl(null, "essentials.jail.notify", tlKey, objects);
```
This feels iffy

JRoy replied:
oh well

JRoy committed:
Placeholder -> ParsedPlaceholder

JRoy committed:
Make conversion method names more consistent

JRoy committed:
Create CommonPlaceholders

JRoy committed:
Okayeg

JRoy committed:
Add log message for adventure conversion

JRoy committed:
Fix helpop behavior change

JRoy committed:
Clarify SignException tlKey parameter

JRoy committed:
clean up worth a lil bit

JRoy committed:
clean up sell a lil bit

JRoy committed:
displayname fuck ah

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandignore.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandseen.java
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandwhois.java

JRoy committed:
fix

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
a reminder

Potothingi commented:
Build failed?

JRoy commented:
> Build failed?

yes, they will fail until all translation strings are converted to mini message. this will be done right before merging

PapiCapi commented:
Hello, is there any news on this or something we could do to help you ?

JRoy commented:
> Hello, is there any news on this or something we could do to help you ?

we're working on finalizing translations and documentation, The code is pretty much complete at this point

mdcfe commented on Essentials/src/main/resources/config.yml:
```diff
@@ -447,6 +447,13 @@ debug: false
 # For more information, visit http://wiki.ess3.net/wiki/Locale
 #locale: en
```
```suggestion
# For more information, visit https://essentialsx.net/wiki/Locale.html
#locale: en
```
Technically separate but ought to be done in one go to match documentation

mdcfe committed:
Update locale docs link in config

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/Settings.java:
```diff
@@ -1948,4 +1962,39 @@ public boolean showZeroBaltop() {
+    }
+
+    private Tag _getSecondaryColor() {
+        final String color = config.getString("primary-color", "c");
```
```suggestion
        final String color = config.getString("message-colors.secondary", "ff5555");
```

mdcfe commented on Essentials/src/main/resources/config.yml:
```diff
@@ -444,9 +444,16 @@ debug: false
+# and EssentialsX will send messages in French to the player and messages in the console as English.
+# If a player's language is not known, the server's language (or one defined above) will be used.
+per-player-locale: false
+
```
```suggestion

# Change the default primary and secondary colours used in EssentialsX messages.
# Some messages may use custom colours, which will need to be edited in the appropriate message files.
# For more information on customising messages, see https://essentialsx.net/wiki/Locale.html
message-colors:
  primary: '#ffaa00'
  secondary: '#ff5555'

```
Pending selection of a better colour palette.  
Using a map allows us toto support other colour aliases here in the future

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/Settings.java:
```diff
@@ -1948,4 +1962,39 @@ public boolean showZeroBaltop() {
+    }
+
+    private Tag _getPrimaryColor() {
+        final String color = config.getString("primary-color", "6");
```
```suggestion
        final String color = config.getString("message-colors.primary", "#ffaa00");
```
We may want to just use hex codes?

JRoy committed:
Apply suggestions from code review

Co-authored-by: MD <1917406+mdcfe@users.noreply.github.com>

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/Settings.java:
```diff
@@ -1948,4 +1962,39 @@ public boolean showZeroBaltop() {
+    }
+
+    private Tag _getSecondaryColor() {
+        final String color = config.getString("message-colors.secondary", "ff5555");
```
```suggestion
        final String color = config.getString("message-colors.secondary", "#ff5555");
```

i have brain cancer

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/Settings.java

JRoy committed:
Support named colors for customizable color codes

JRoy committed:
Support customizable colors for the upgrade

JRoy committed:
Enable the upgrade

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/we-are-going-on-an-adventure

TheKrafter commented:
Just your friendly neighborhood server admin, I was wondering if we could get a progress report of some kind. I really appreciate yall's work, and also had a few questions regarding these changes that the answers to are probably well-known to EssentialsX developers, but I have not been able find answers to elsewhere:
- Does this PR add PlaceholderAPI Support? (as #2041 started, then was closed in favor of #2502, which was closed in favor of this PR)
- Does this PR add MiniPlaceholder support?
- Would the `{PLACEHOLDER}` style placeholders in configs and `*.txt` files need to be modified?

I don't want to be a nag, but I really do appreciate your work and would like to know how this all is going so far, thank you all very much!

mdcfe committed:
Update Essentials/src/main/java/com/earth2me/essentials/Essentials.java

Co-authored-by: pop4959 <pop4959@gmail.com>

JRoy committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

JRoy committed:
Bump Adventure

JRoy committed:
Move custom message files to separate directory

JRoy committed:
Fix minor adventure trolling

JRoy committed:
Preserve messages order

JRoy committed:
Add some javadocs

people may forget what these things do later

codebycam commented:
> LGTM

Still LGTM

mdcfe committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

mdcfe committed:
Merge branch '2.x' into refactor/we-are-going-on-an-adventure

mdcfe committed:
Update messages files to use MiniMessage

mdcfe committed:
Set up BukkitAudiences in unit tests

SrBedrock commented:
When a player joins the server, an error is sent: https://pastes.dev/YbBHLlf2fw
When trying to receive messages from Essentials an error is sent: https://pastes.dev/lB3llBOiDM

![image](https://github.com/EssentialsX/Essentials/assets/51332006/9963a740-c7d3-42c1-8b27-ab2fc6306372)

mbax commented:
> When a player joins the server, an error is sent: https://pastes.dev/YbBHLlf2fw When trying to receive messages from Essentials an error is sent: https://pastes.dev/lB3llBOiDM
> 
> ![image](https://private-user-images.githubusercontent.com/51332006/302057380-9963a740-c7d3-42c1-8b27-ab2fc6306372.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY5OTEwODcsIm5iZiI6MTcwNjk5MDc4NywicGF0aCI6Ii81MTMzMjAwNi8zMDIwNTczODAtOTk2M2E3NDAtYzdkMy00MmMxLThiMjctYWIyZmM2MzA2MzcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAzVDIwMDYyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkOTk0NDFhOTRkNjNkY2UxNjhjNDNmZjg2ZjgwZWY1MDFjOWNkYWI4NmJiYWU4ZGQ1ZGRjODA1NzdkZTUyNDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SDzV03NMbYnHN4THMyPX8v8H0l97LMwPqeLFtIW9GOU)

Any chance you were live swapping the EssX jar in the `plugins` folder instead of using `plugins/update` ?

mdcfe committed:
Shade kyori option lib

mdcfe commented:
@SrBedrock The PR has been updated to shade the missing options library - would you be able to update to the latest dev build?
